### DISCUSSION TITLE
Support for different hosting platform workflows

### DISCUSSION THREAD

**Discussion by ChrisJStone** [CJ-Systems/gitflow-cjs#44] [Ideas]
Since git flow could be considered a devops tool, it should provide support for the different code hosting platforms that are available. While there is support for interacting with remote repositories there is no clearly defined process in place on how to work with online platforms such as Github, Gitlab and the like.

I can currently see 4 possible solutions to this:

1. Since git flow all ready is a universal tool we do nothing and leave it up to the end user to figure out how to implement git flow with the platform being used.
2. Define and implement minimal functionality that will work with most platforms.
3. Implement specific support for each available platform.
4. A combination of the above either through the use of platform specific support, common functionality or available tools/plugins such as the gh cli for Github.

Out of the available solutions I believe the 4th option would be the best way to proceed I feel this will have the minimal number of drawbacks as a whole since as part of the implementation we can define and implement minimal viable functionality as suggested in solution 2 and be able to release a initial update to git flow.

The drawbacks to the solution I can see are:

1. Someone new to git flow may get the impression that it will not work with the platform they are using either due to it not being specifically supported or a feature of the platform they use is not part of the commonly defined functionality.
2. Existing workflows where git flow is implemented break as a result of the updated feature set.
3. Additional external dependencies needed just to get started using git flow.

In order to mitigate the impact of the above drawbacks proper documentation on git flow beyond what is currently covered in the wiki must be created. Secondly while some breakage is expected to happen we can provide LTS releases that only include updates that are backwards compatible with the existing feature set. Finally while the use of external dependencies may be necessary if at all possible they should be optional and if not properly document why they are necessary for the use of git flow.

Please use this discussion thread for general comments and ideas related to the above proposal or for general comments related to a proposed SRS or RFC. If you wish to submit a workflow for a specific platform such as Github  please create a topic for it in SRS Proposals and link it here with a short summary describing the workflow. For changes to an existing workflow please use the RFC Proposals topic instead.

**Comment by ChrisJStone** [comment]
For RFC Proposals reply to this comment

**Reply by ChrisJStone** [reply]
#46 Support Require signed commits on github protected branches

This RFC still needs fleshing out.

**Reply by ChrisJStone** [reply]
#47 Address git feature pull/track/checkout

**Comment by ChrisJStone** [comment]
FOR SRS Proposals reply to this comment

**Comment by shaedrich** [comment]
What about making gitflow "extendable" so that its core doesn't have to include it by default but instead offers a command with which you can easily install platform-specific functionality as well as maybe some related tools?

**Reply by ChrisJStone** [reply]
I think that might have been the idea behind the hooks. I do like the idea of having git flow "extendable" it is something I've thought about in the past. Granted it's on a lot larger scale and deals with more then just a branching workflow. When I was first thinking how to solve the integration issues with github I had considered adding a config option where if enabled the gh cli would be used as much as possible. For example when running feature finish instead of git being called gh would push the feature branch, create a pr, merge the pr and close it out then pull develop down.

**Reply by shaedrich** [reply]
I love the idea to choose your git platform during configuration and then fall back on platform-native functionality via proprietary command line programs that might even offer their own hooks or the like.

**Reply by ChrisJStone** [reply]
Agreed, I believe this will also help to mitigate the first drawback I mentioned as this would allow for a  list the user could select from as an example:

Choose  hosting platform:
1) Github
2) Gitlab
3) Bitbucket
4) Other use native git flow scripts

Or something along that line.

**Reply by shaedrich** [reply]
Yep, sounds good 👍🏻
